aquadoggo/src/materializer/tasks/reduce.rs

```diff
@@ -1,17 +1,242 @@
+                .store
+                .get_document_by_operation_id(operation_id)
+                .await
+                .map_err(|_| TaskError::Critical)? // We only get an error here on a critical database error.
+            {
+                Some(document_id) => Ok(document_id),
+                None => Err(TaskError::Failure),
+            }
+        }
+        (_, _) => Err(TaskError::Failure),
+    }?;
 
-    // @TODO: Store document view in database
-    Ok(None)
+    let operations = context
+        .store
+        .get_operations_by_document_id(&document_id)
+        .await
+        .map_err(|_| TaskError::Critical)?
+        .into_iter()
+        // TODO: we can avoid this conversion if we do https://github.com/p2panda/p2panda/issues/320
+        .map(|op| op.into())
+        .collect();
+
+    let document_view_id = match &input.document_view_id {
+        Some(document_view_id) => {
+            // TODO: If we are resolving a document_view_id, but we are missing operations from it's graph,
+            // what do we want to return?
+            let document = DocumentBuilder::new(operations)
+                .build_to_view_id(Some(document_view_id.to_owned()))
+                .map_err(|_| TaskError::Critical)?;
+
+            if document.is_deleted() {
+                return Ok(None);
+            } else {
+                context
+                    .store
+                    // Unwrap as all not deleted documents have a view.
+                    .insert_document_view(document.view().unwrap(), document.schema())
+                    .await
+                    .map_err(|_| TaskError::Critical)?;
+
+                document_view_id.to_owned()
+            }
+        }
+        None => {
+            let document = DocumentBuilder::new(operations)
+                .build()
+                .map_err(|_| TaskError::Failure)?;
+
+            context
+                .store
+                .insert_document(&document)
+                .await
+                .map_err(|_| TaskError::Failure)?;
+
+            if document.is_deleted() {
+                return Ok(None);
+            } else {
+                document.view_id().to_owned()
+            }
+        }
+    };
+
+    Ok(Some(vec![Task::new(
+        "dependency",
+        TaskInput::new(None, Some(document_view_id)),
+    )]))
+}
+
+#[cfg(test)]
+mod tests {
+    use p2panda_rs::document::{DocumentId, DocumentViewId};
+    use p2panda_rs::operation::OperationValue;
+    use p2panda_rs::storage_provider::traits::{AsStorageOperation, OperationStore};
+    use p2panda_rs::test_utils::constants::TEST_SCHEMA_ID;
+    use p2panda_rs::test_utils::fixtures::{random_document_id, random_document_view_id};
+    use rstest::rstest;
+
+    use crate::config::Configuration;
+    use crate::context::Context;
+    use crate::db::stores::test_utils::{test_db, TestSqlStore};
+    use crate::db::traits::DocumentStore;
+    use crate::materializer::tasks::reduce_task;
```

Comment:
This you can import via `super`, then it's easy to see what objects were testing in that file 👍